homeassistant/components/group/__init__.py

```diff
@@ -293,14 +294,39 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     await _async_process_config(hass, config)
 
     async def reload_service_handler(service: ServiceCall) -> None:
-        """Remove all user-defined groups and load new ones from config."""
-        auto = [e for e in component.entities if not e.user_defined]
+        """Group reload handler.
 
-        if (conf := await component.async_prepare_reload()) is None:
+        - Remove group.group entities not created by service calls and set them up again
+        - Reload xxx.group platforms
+        """
+        if (conf := await component.async_prepare_reload(skip_reset=True)) is None:
             return
-        await _async_process_config(hass, conf)
 
-        await component.async_add_entities(auto)
+        async def reset_group_platform(platform: EntityPlatform) -> None:
+            # Simplified + modified version of EntityPlatform.async_reset:
+            # - group.group never retries setup
+            # - group.group never polls
+            # - We don't need to reset EntityPlatform._setup_complete
+            # - Only remove entities which were not created by service calls
+            if not platform.entities:
+                return
+
+            tasks = [
+                entity.async_remove()
+                for entity in platform.entities.values()
+                if cast(Group, entity).user_defined
+            ]
+
+            await asyncio.gather(*tasks)
+
```

Comment:
This this still need to be an inner function?

Reply:
Not really, do you think it's clear if it's not an inner function?

Reply:
Yes. I generally prefer to avoid inner functions unless we need to capture a variable since they are harder to follow the flow since you have to mentally back track and keep another state in the brain while keeping track of the flow